developer/services/qrexec3.md

```diff
@@ -16,94 +16,79 @@ redirect_from:
+The qrexec framework generalizes this process.
+It allows users and developers to use and design secure inter-VM tools.
+
+## Qrexec basics: architecture and examples ##
+
+Qrexec is built on top of *vchan*, a Xen library providing data links between VMs.
+During domain creation, a process named `qrexec-daemon` is started in dom0, and a process named `qrexec-agent` is started in the VM.
+They are connected over a **vchan** channel.
+`qrexec-daemon` listens for connections from a dom0 utility named `qrexec-client`.
+Let's say we want to start a process (call it `VMprocess`) in a VM (`someVM`).
+Typically, the first thing that a `qrexec-client` instance does is to send a request to the `qrexec-daemon`, which in turn relays it to `qrexec-agent` running in `someVM`.
+`qrexec-daemon` assigns unique vchan connection details and sends them to both `qrexec-client` (in dom0) and `qrexec-agent` (in `someVM`).
+`qrexec-client` starts a vchan server, which `qrexec-agent` then connects to.
+Once this channel is established, stdin/stdout/stderr from the VMprocess is passed between `qrexec-agent` and the `qrexec-client` process.
+
+The `qrexec-client` command is used to make connections to VMs from dom0.
+For example, the following command
 
-allows to work with the remote shell. The string before the first
-semicolon specifies what user to run the command as. Adding `-e` on the
-`qrexec-client` command line results in mere command execution (no data
-passing), and `qrexec-client` exits immediately after sending the execution
-request and receiving status code from `qrexec-agent` (whether the process
-creation succeeded). There is also the `-l local_program` flag -- with it,
-`qrexec-client` passes stdin/stdout of the remote process to the (spawned
-for this purpose) `local_program`, not to its own stdin/stdout.
+    qrexec-client -e -d someVM user:'touch hello-world.txt'
 
-The `qvm-run` command is heavily based on `qrexec-client`. It also takes care
-of additional activities, e.g. starting the domain if it is not up yet and
-starting the GUI daemon. Thus, it is usually more convenient to use `qvm-run`.
+creates an empty file called `hello-world.txt` in the home folder of `someVM`.
 
-There can be almost arbitrary number of `qrexec-client` processes for a
-domain (so, connected to the same `qrexec-daemon`, same domain) -- their
-data is multiplexed independently. Number of available vchan channels is
-the limiting factor here, it depends on the underlying hypervisor.
+The string before the colon specifies what user to run the command as.
+The `-e` flag tells `qrexec-client` to exit immediately after sending the execution request and receiving a status code from `qrexec-agent` (whether the process creation succeeded).
+With this option, no further data is passed between the domains.
+By contrast, the following command demonstrates an open channel between two VMs: in this case, a remote shell.
```

Comment:
"between two VMs" - since in this example one end is dom0, this may not be fully accurate. Actually, very recently there was discussion on xen-devel ML whether dom0 is a VM (for the purpose of setting up glossary wiki page). Consensus wasn't reached yet...